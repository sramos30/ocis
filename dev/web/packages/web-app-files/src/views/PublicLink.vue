<template>
  <div class="oc-height-1-1 oc-link-resolve">
    <h1 class="oc-invisible-sr">{{ pageTitle }}</h1>
    <div class="oc-card oc-border oc-rounded oc-position-center oc-text-center oc-width-large">
      <template v-if="loading">
        <div class="oc-card-header">
          <h2 key="public-link-loading">
            <translate>Loading public linkâ€¦</translate>
          </h2>
        </div>
        <div class="oc-card-body">
          <oc-spinner :aria-hidden="true" />
        </div>
      </template>
      <template v-else-if="errorMessage">
        <div class="oc-card-header oc-link-resolve-error-title">
          <h2 key="public-link-error">
            <translate>An error occurred while loading the public link</translate>
          </h2>
        </div>
        <div class="oc-card-body oc-link-resolve-error-message">
          <p class="oc-text-lead">{{ errorMessage }}</p>
        </div>
      </template>
      <template v-else-if="passwordRequired">
        <form @submit.prevent="resolvePublicLink">
          <div class="oc-card-header">
            <h2>
              <translate>This resource is password-protected</translate>
            </h2>
          </div>
          <div class="oc-card-body">
            <oc-text-input
              ref="passwordInput"
              v-model="password"
              :error-message="inputErrorMessage"
              :label="passwordFieldLabel"
              type="password"
              class="oc-mb-s"
            />
            <oc-button
              variation="primary"
              appearance="filled"
              class="oc-login-authorize-button"
              :disabled="!password"
              submit="submit"
            >
              <translate>Continue</translate>
            </oc-button>
          </div>
        </form>
      </template>
      <div class="oc-card-footer">
        <p>
          {{ configuration.currentTheme.general.slogan }}
        </p>
      </div>
    </div>
  </div>
</template>

<script>
import { mapGetters, mapActions } from 'vuex'
import Mixins from '../mixins'
import { DavProperties } from 'web-pkg/src/constants'
import { createLocationPublic } from '../router'

export default {
  mixins: [Mixins],
  data() {
    return {
      loading: true,
      errorMessage: null,
      firstTime: true,
      passwordRequired: false,
      password: null,
      inputErrorMessage: null
    }
  },
  computed: {
    ...mapGetters(['configuration']),
    ...mapGetters('Files', ['publicLinkPassword']),

    pageTitle() {
      return this.$gettext(this.$route.meta.title)
    },

    passwordFieldLabel() {
      return this.$gettext('Enter password for public link')
    }
  },
  created() {
    this.resolvePublicLink()
  },
  methods: {
    ...mapActions('Files', ['setPublicLinkPassword']),
    resolvePublicLink() {
      const password = this.password || this.publicLinkPassword
      this.loading = true
      this.inputErrorMessage = null
      this.$client.publicFiles
        .list(this.$route.params.token, password, DavProperties.PublicLink, '0')
        .then((files) => {
          this.passwordRequired = false
          this.setPublicLinkPassword(password)
          const publicOperationName =
            files[0].getProperty(this.$client.publicFiles.PUBLIC_LINK_PERMISSION) === '4'
              ? 'files-public-drop'
              : 'files-public-files'

          this.$router.push(
            createLocationPublic(publicOperationName, {
              params: {
                item: this.$route.params.token
              }
            })
          )
        })
        .catch((error) => {
          if (error.statusCode === 401) {
            this.passwordRequired = true
            if (!this.firstTime) {
              this.inputErrorMessage = this.$gettext('Incorrect password')
            }
            this.firstTime = false
            // somehow needed here for the focus call to work correctly,
            // otherwise the element is not rendered yet
            this.loading = false
            this.$nextTick(() => {
              this.$refs.passwordInput.focus()
            })
          } else {
            this.errorMessage = error
          }
        })
        .finally(() => {
          this.loading = false
        })
    }
  }
}
</script>

<style lang="scss">
.oc-link-resolve {
  .oc-text-input-message {
    justify-content: center;
  }

  .oc-card-header h2,
  .oc-card-footer p {
    margin: 0;
  }
}
</style>
