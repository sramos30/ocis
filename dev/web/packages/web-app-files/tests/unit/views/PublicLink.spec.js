import PublicLink from '@files/src/views/PublicLink.vue'
import { getStore, localVue } from './views.setup.js'
import { shallowMount, mount } from '@vue/test-utils'

const theme = {
  general: { slogan: 'some-slogan' }
}

const stubs = {
  'oc-spinner': true,
  translate: true
}

const route = { meta: { title: 'some page title' }, params: { token: 'some-token' } }

const component = { ...PublicLink, mounted: jest.fn(), created: jest.fn() }

const selectors = {
  cardFooter: '.oc-card-footer',
  cardBody: '.oc-card-body',
  submitButton: '.oc-login-authorize-button',
  errorTitle: '.oc-link-resolve-error-title',
  publicLinkPasswordRequired: 'form > .oc-card-title'
}

const spinnerStub = 'oc-spinner-stub'
const textInputStub = 'oc-text-input-stub'

describe('PublicLink', () => {
  describe('theming options', () => {
    const wrapper = getWrapper()

    it('should have the page-title set', () => {
      expect(wrapper).toMatchSnapshot()
    })
    it('should display the configuration theme general slogan as the login card footer', () => {
      const slogan = wrapper.find(selectors.cardFooter)

      expect(slogan).toMatchSnapshot()
    })
  })

  describe('when the view is still loading', () => {
    const wrapper = getWrapper({ loading: true })
    it('should display the loading text with the spinner', () => {
      const loading = wrapper.find(selectors.cardBody)

      expect(wrapper.find(spinnerStub).exists()).toBeTruthy()
      expect(loading).toMatchSnapshot()
    })
    it('should not display the error message', () => {
      expect(wrapper.find(selectors.errorTitle).exists()).toBeFalsy()
    })
    it('should not display the password required form', () => {
      expect(wrapper.find(selectors.publicLinkPasswordRequired).exists()).toBeFalsy()
    })
  })

  describe('when the view is not loading anymore', () => {
    it('should not display the loading text and the spinner', () => {
      const wrapper = getWrapper({ loading: false })
      const loading = wrapper.find(selectors.cardBody)

      expect(loading.exists()).toBeFalsy()
    })
    it('should display the error message if "errorMessage" is not empty', async () => {
      const wrapper = getWrapper({ loading: false })
      await wrapper.setData({ errorMessage: 'some-error-message' })

      expect(wrapper.find(selectors.errorTitle).exists()).toBeTruthy()
      expect(wrapper.find(selectors.errorTitle)).toMatchSnapshot()
    })

    describe('and when "passwordRequired" is set as true', () => {
      let passwordRequiredForm
      const wrapper = getWrapper({ loading: false })

      beforeEach(async () => {
        await wrapper.setData({ passwordRequired: true })
        passwordRequiredForm = wrapper.find('form')
      })

      it('should display the password required form', () => {
        expect(passwordRequiredForm.exists()).toBeTruthy()
        expect(passwordRequiredForm).toMatchSnapshot()
      })

      describe('and the password input', () => {
        it('should have a computed label', () => {
          const passwordInput = passwordRequiredForm.find(textInputStub)

          expect(passwordInput).toMatchSnapshot()
        })
        it('should not display the error message if "inputErrorMessage" is falsy', () => {
          const passwordInput = passwordRequiredForm.find(textInputStub)

          expect(passwordInput.attributes().errormessage).toBeUndefined()
        })
        it('should display the error message if "inputErrorMessage" is not empty', async () => {
          await wrapper.setData({ inputErrorMessage: 'some input error message' })
          const passwordInput = passwordRequiredForm.find(textInputStub)

          expect(passwordInput).toMatchSnapshot()
        })
      })

      describe('and the submit button', () => {
        it('should be set as disabled if "password" is empty', () => {
          expect(wrapper.vm.password).toBeNull()

          const submitButton = wrapper.find(selectors.submitButton)
          expect(submitButton.exists()).toBeTruthy()
          expect(submitButton).toMatchSnapshot()
        })
        it('should be set as enabled if "password" is not empty', async () => {
          await wrapper.setData({ password: 'some-value' })
          expect(wrapper.vm.password).toBe('some-value')

          const submitButton = wrapper.find(selectors.submitButton)
          expect(submitButton.exists()).toBeTruthy()
          expect(submitButton.attributes().disabled).toBeUndefined()
        })
      })

      it('should call "resolvePublicLink" method on form submit', async () => {
        const spyResolvePublicLink = jest.spyOn(PublicLink.methods, 'resolvePublicLink')
        const wrapper = getMountedWrapper({ loading: false })

        await wrapper.setData({ passwordRequired: true, password: 'some-pass' })

        const submitButton = wrapper.find(selectors.submitButton)
        await submitButton.trigger('submit')

        expect(spyResolvePublicLink).toHaveBeenCalledTimes(1)
      })
    })
  })
})

function getMountOptions({
  $route = route,
  store = getStore({
    slogan: theme.general.slogan
  }),
  loading = false
} = {}) {
  return {
    localVue,
    store,
    stubs,
    mocks: {
      $route
    },
    data: () => ({
      loading: loading
    })
  }
}

function getWrapper({ loading, store } = {}) {
  return shallowMount(component, getMountOptions({ store, loading }))
}

function getMountedWrapper({ loading, store } = {}) {
  return mount(component, {
    ...getMountOptions({ store, loading }),
    stubs: {}
  })
}
