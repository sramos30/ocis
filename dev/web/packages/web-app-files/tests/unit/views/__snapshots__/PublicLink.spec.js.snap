// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PublicLink theming options should display the configuration theme general slogan as the login card footer 1`] = `
<div class="oc-card-footer">
  <p>
    some-slogan
  </p>
</div>
`;

exports[`PublicLink theming options should have the page-title set 1`] = `
<div class="oc-height-1-1 oc-link-resolve">
  <h1 class="oc-invisible-sr">some page title</h1>
  <div class="oc-card oc-border oc-rounded oc-position-center oc-text-center oc-width-large">
    <!---->
    <div class="oc-card-footer">
      <p>
        some-slogan
      </p>
    </div>
  </div>
</div>
`;

exports[`PublicLink when the view is not loading anymore and when "passwordRequired" is set as true and the password input should display the error message if "inputErrorMessage" is not empty 1`] = `<oc-text-input-stub id="oc-textinput-3" type="password" clearbuttonaccessiblelabel="" label="Enter password for public link" class="oc-mb-s" errormessage="some input error message"></oc-text-input-stub>`;

exports[`PublicLink when the view is not loading anymore and when "passwordRequired" is set as true and the password input should have a computed label 1`] = `<oc-text-input-stub id="oc-textinput-3" type="password" clearbuttonaccessiblelabel="" label="Enter password for public link" class="oc-mb-s"></oc-text-input-stub>`;

exports[`PublicLink when the view is not loading anymore and when "passwordRequired" is set as true and the submit button should be set as disabled if "password" is empty 1`] = `
<oc-button-stub type="button" disabled="true" size="medium" submit="submit" variation="primary" appearance="filled" justifycontent="center" gapsize="medium" class="oc-login-authorize-button">
  <translate-stub tag="span">Continue</translate-stub>
</oc-button-stub>
`;

exports[`PublicLink when the view is not loading anymore and when "passwordRequired" is set as true should display the password required form 1`] = `
<form>
  <div class="oc-card-header">
    <h2>
      <translate-stub tag="span">This resource is password-protected</translate-stub>
    </h2>
  </div>
  <div class="oc-card-body">
    <oc-text-input-stub id="oc-textinput-3" type="password" clearbuttonaccessiblelabel="" label="Enter password for public link" class="oc-mb-s"></oc-text-input-stub>
    <oc-button-stub type="button" disabled="true" size="medium" submit="submit" variation="primary" appearance="filled" justifycontent="center" gapsize="medium" class="oc-login-authorize-button">
      <translate-stub tag="span">Continue</translate-stub>
    </oc-button-stub>
  </div>
</form>
`;

exports[`PublicLink when the view is not loading anymore should display the error message if "errorMessage" is not empty 1`] = `
<div class="oc-card-header oc-link-resolve-error-title">
  <h2>
    <translate-stub tag="span">An error occurred while loading the public link</translate-stub>
  </h2>
</div>
`;

exports[`PublicLink when the view is still loading should display the loading text with the spinner 1`] = `
<div class="oc-card-body">
  <oc-spinner-stub arialabel="" size="medium" aria-hidden="true"></oc-spinner-stub>
</div>
`;
